File: src/test/java/dev/retrotv/crypt/owe/OWETest.java
package dev.retrotv.crypt.owe;

import dev.retrotv.common.Log;
import dev.retrotv.crypt.owe.crc.CRC32;
import dev.retrotv.crypt.owe.md.*;
import dev.retrotv.crypt.owe.sha.*;
import dev.retrotv.enums.HashAlgorithm;
import org.json.JSONObject;

import java.io.*;
import java.net.URISyntaxException;
import java.net.URL;
import java.nio.file.Files;
import java.util.Objects;

import static org.junit.jupiter.api.Assertions.*;

public class OWETest extends Log {
    protected final String PASSWORD = "The quick brown fox jumps over the lazy dog";
    protected  final URL CHECKSUM = this.getClass().getClassLoader().getResource("checksum");
    protected final URL RESOURCE = this.getClass().getClassLoader().getResource("checksum_test_file.txt");
    protected final URL RESOURCE2 = this.getClass().getClassLoader().getResource("checksum_test_file2.txt");

    protected void fileHashTest(HashAlgorithm algorithm) throws IOException {
        File file;
        byte[] fileData;

        try {
            file = new File(Objects.requireNonNull(RESOURCE).toURI());
        } catch (URISyntaxException e) {
            throw new RuntimeException(e);
        }

        try (DataInputStream dis = new DataInputStream(Files.newInputStream(file.toPath()))) {
            fileData = new byte[(int) file.length()];
            dis.readFully(fileData);
        } catch (IOException e) {
            throw new IOException("파일을 읽어들이는 과정에서 예상치 못한 오류가 발생했습니다.");
        }

        assertEquals(getHash(algorithm), hash(algorithm, fileData));
    }

    protected void fileHashMatchesTest(Checksum checksum, HashAlgorithm algorithm) throws IOException {
        File file;
        byte[] fileData;

        try {
            file = new File(Objects.requireNonNull(RESOURCE).toURI());
        } catch (URISyntaxException e) {
            throw new RuntimeException(e);
        }

        try (DataInputStream dis = new DataInputStream(Files.newInputStream(file.toPath()))) {
            fileData = new byte[(int) file.length()];
            dis.readFully(fileData);
        } catch (IOException e) {
            throw new IOException("파일을 읽어들이는 과정에서 예상치 못한 오류가 발생했습니다.");
        }

        assertTrue(checksum.matches(fileData, getHash(algorithm)));
    }

    protected void fileMatchesTest(Checksum checksum) throws IOException {
        if (RESOURCE != null && RESOURCE2 != null) {
            assertTrue(checksum.matches(new File(RESOURCE.getFile()), new File(RESOURCE2.getFile())));
        } else {
            fail();
        }
    }

    protected void passwordEncryptAndMatchesTest(Password password) {
        String encryptedPassword = password.encode(PASSWORD);

        log.info(encryptedPassword);

        assertNotEquals(PASSWORD, encryptedPassword);
        assertTrue(password.matches(PASSWORD, encryptedPassword));
    }

    private String hash(HashAlgorithm algorithm, byte[] fileData) {
        switch (algorithm) {
            case CRC32: {
                Checksum checksum = new CRC32();
                return checksum.hash(fileData);
            }

            case MD2: {
                Checksum checksum = new MD2();
                return checksum.hash(fileData);
            }

            case MD5: {
                Checksum checksum = new MD5();
                return checksum.hash(fileData);
            }

            case SHA1:  {
                Checksum checksum = new SHA1();
                return checksum.hash(fileData);
            }

            case SHA224: {
                Checksum checksum = new SHA224();
                return checksum.hash(fileData);
            }

            case SHA256:  {
                Checksum checksum = new SHA256();
                return checksum.hash(fileData);
            }

            case SHA384: {
                Checksum checksum = new SHA384();
                return checksum.hash(fileData);
            }

            case SHA512: {
                Checksum checksum = new SHA512();
                return checksum.hash(fileData);
            }

            case SHA512224: {
                Checksum checksum = new SHA512224();
                return checksum.hash(fileData);
            }

            case SHA512256: {
                Checksum checksum = new SHA512256();
                return checksum.hash(fileData);
            }

            default: return null;
        }
    }

    private String getHash(HashAlgorithm algorithm) throws IOException {
        JSONObject jsonObject = new JSONObject(readJson());
        JSONObject file1 = jsonObject.getJSONObject("checksum_test_file");

        switch (algorithm) {
            case CRC32: return file1.getString(HashAlgorithm.CRC32.label());
            case MD2: return file1.getString(HashAlgorithm.MD2.label());
            case MD5: return file1.getString(HashAlgorithm.MD5.label());
            case SHA1: return file1.getString(HashAlgorithm.SHA1.label());
            case SHA224: return file1.getString(HashAlgorithm.SHA224.label());
            case SHA256: return file1.getString(HashAlgorithm.SHA256.label());
            case SHA384: return file1.getString(HashAlgorithm.SHA384.label());
            case SHA512: return file1.getString(HashAlgorithm.SHA512.label());
            case SHA512224: return file1.getString(HashAlgorithm.SHA512224.label());
            case SHA512256: return file1.getString(HashAlgorithm.SHA512256.label());
            default: return null;
        }
    }

    private String readJson() throws IOException {
        if (CHECKSUM == null) {
            throw new IOException();
        }

        String json;

        try(BufferedReader reader = new BufferedReader(new FileReader(CHECKSUM.getFile()))) {
            StringBuilder sb = new StringBuilder();
            String line = reader.readLine();

            while (line != null) {
                sb.append(line);
                sb.append("\n");
                line = reader.readLine();
            }
            json = sb.toString();
        }

        log.info(json);

        return json;
    }
}

File: src/test/kotlin/dev/retrotv/crypto/owe/OWETest.kt
package dev.retrotv.crypto.owe

import dev.retrotv.common.Log
import dev.retrotv.crypto.owe.hash.Hash
import dev.retrotv.crypto.owe.hash.crc.CRC32
import dev.retrotv.crypto.owe.hash.md.*
import dev.retrotv.crypto.owe.hash.sha.*
import dev.retrotv.enums.Algorithm
import org.json.JSONObject
import org.junit.jupiter.api.Assertions
import org.springframework.security.crypto.password.PasswordEncoder
import java.io.*
import java.net.URISyntaxException
import java.nio.file.Files
import java.util.*

open class OWETest : Log() {
    private val PASSWORD = "The quick brown fox jumps over the lazy dog"
    private val CHECKSUM = this.javaClass.getClassLoader().getResource("hash_code")
    private val RESOURCE = this.javaClass.getClassLoader().getResource("hash_code_test_file.txt")

    @Throws(IOException::class)
    protected fun fileHashTest(algorithm: Algorithm.Hash) {
        var fileData: ByteArray
        val file = try {
            File(Objects.requireNonNull(RESOURCE).toURI())
        } catch (e: URISyntaxException) {
            throw RuntimeException(e)
        }

        try {
            DataInputStream(Files.newInputStream(file.toPath())).use { dis ->
                fileData = ByteArray(file.length().toInt())
                dis.readFully(fileData)
            }
        } catch (e: IOException) {
            throw IOException("파일을 읽어들이는 과정에서 예상치 못한 오류가 발생했습니다.")
        }

        println(getHash(algorithm))
        println(hash(algorithm, fileData))

        Assertions.assertEquals(getHash(algorithm), hash(algorithm, fileData))
    }

    @Throws(IOException::class)
    protected fun fileHashMatchesTest(fileHash: Hash, algorithm: Algorithm.Hash) {
        var fileData: ByteArray
        val file = try {
            File(Objects.requireNonNull(RESOURCE).toURI())
        } catch (e: URISyntaxException) {
            throw RuntimeException(e)
        }

        try {
            DataInputStream(Files.newInputStream(file.toPath())).use { dis ->
                fileData = ByteArray(file.length().toInt())
                dis.readFully(fileData)
            }
        } catch (e: IOException) {
            throw IOException("파일을 읽어들이는 과정에서 예상치 못한 오류가 발생했습니다.")
        }

        Assertions.assertTrue(fileHash.matches(fileData, getHash(algorithm)))
    }

    protected fun passwordEncryptAndMatchesTest(password: PasswordEncoder) {
        val encryptedPassword = password.encode(PASSWORD)
        log.info(encryptedPassword)

        Assertions.assertNotEquals(PASSWORD, encryptedPassword)
        Assertions.assertTrue(password.matches(PASSWORD, encryptedPassword))
    }

    private fun hash(algorithm: Algorithm.Hash, fileData: ByteArray): String? {
        return when (algorithm) {
            Algorithm.Hash.CRC32 -> {
                val hash: Hash = CRC32()
                hash.hash(fileData)
            }

            Algorithm.Hash.MD2 -> {
                val hash: Hash = MD2()
                hash.hash(fileData)
            }

            Algorithm.Hash.MD5 -> {
                val hash: Hash = MD5()
                hash.hash(fileData)
            }

            Algorithm.Hash.SHA1 -> {
                val hash: Hash = SHA1()
                hash.hash(fileData)
            }

            Algorithm.Hash.SHA224 -> {
                val hash: Hash = SHA224()
                hash.hash(fileData)
            }

            Algorithm.Hash.SHA256 -> {
                val hash: Hash = SHA256()
                hash.hash(fileData)
            }

            Algorithm.Hash.SHA384 -> {
                val hash: Hash = SHA384()
                hash.hash(fileData)
            }

            Algorithm.Hash.SHA512 -> {
                val hash: Hash = SHA512()
                hash.hash(fileData)
            }

            Algorithm.Hash.SHA512224 -> {
                val hash: Hash = SHA512224()
                hash.hash(fileData)
            }

            Algorithm.Hash.SHA512256 -> {
                val hash: Hash = SHA512256()
                hash.hash(fileData)
            }

            Algorithm.Hash.SHA3224 -> {
                val hash: Hash = SHA3224()
                hash.hash(fileData)
            }

            Algorithm.Hash.SHA3256 -> {
                val hash: Hash = SHA3256()
                hash.hash(fileData)
            }

            Algorithm.Hash.SHA3384 -> {
                val hash: Hash = SHA3384()
                hash.hash(fileData)
            }

            Algorithm.Hash.SHA3512 -> {
                val hash: Hash = SHA3512()
                hash.hash(fileData)
            }

            else -> null
        }
    }

    @Throws(IOException::class)
    private fun getHash(algorithm: Algorithm.Hash): String? {
        val jsonObject = JSONObject(readJson())
        val file1 = jsonObject.getJSONObject("hash_code_test_file")
        return when (algorithm) {
            Algorithm.Hash.CRC32 -> file1.getString(Algorithm.Hash.CRC32.label())
            Algorithm.Hash.MD2 -> file1.getString(Algorithm.Hash.MD2.label())
            Algorithm.Hash.MD5 -> file1.getString(Algorithm.Hash.MD5.label())
            Algorithm.Hash.SHA1 -> file1.getString(Algorithm.Hash.SHA1.label())
            Algorithm.Hash.SHA224 -> file1.getString(Algorithm.Hash.SHA224.label())
            Algorithm.Hash.SHA256 -> file1.getString(Algorithm.Hash.SHA256.label())
            Algorithm.Hash.SHA384 -> file1.getString(Algorithm.Hash.SHA384.label())
            Algorithm.Hash.SHA512 -> file1.getString(Algorithm.Hash.SHA512.label())
            Algorithm.Hash.SHA512224 -> file1.getString(Algorithm.Hash.SHA512224.label())
            Algorithm.Hash.SHA512256 -> file1.getString(Algorithm.Hash.SHA512256.label())
            Algorithm.Hash.SHA3224 -> file1.getString(Algorithm.Hash.SHA3224.label())
            Algorithm.Hash.SHA3256 -> file1.getString(Algorithm.Hash.SHA3256.label())
            Algorithm.Hash.SHA3384 -> file1.getString(Algorithm.Hash.SHA3384.label())
            Algorithm.Hash.SHA3512 -> file1.getString(Algorithm.Hash.SHA3512.label())
            else -> null
        }
    }

    @Throws(IOException::class)
    private fun readJson(): String {
        if (CHECKSUM == null) {
            throw IOException()
        }

        var json: String
        BufferedReader(FileReader(CHECKSUM.file)).use { reader ->
            val sb = StringBuilder()
            var line = reader.readLine()
            while (line != null) {
                sb.append(line)
                sb.append("\n")
                line = reader.readLine()
            }
            json = sb.toString()
        }

        log.info(json)
        return json
    }
}
